material-ui muiThemeable decorator not working

I'm having problems using `muiThemeable` decorator, definition below:

``` ts
// material-ui.d.ts
export function muiThemeable<TComponent extends React.Component<P, S>, P, S>(): (component: TComponent) => TComponent;
```
## Attempts

Trying it without generics specified:

``` ts
@muiThemeable()
export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {}
```

Gives:

```
error TS2345: Argument of type 'typeof LoginForm' is not assignable to parameter of type 'Component<{}, {}>'.
  Property 'setState' is missing in type 'typeof LoginForm'.
```

---

With generics specified:

``` ts
@muiThemeable<LoginForm, LoginFormProps, LoginFormState>()
export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {
  private: formNode;
}
```

Gives:

```
error TS2345: Argument of type 'typeof LoginForm' is not assignable to parameter of type 'LoginForm'.
  Property 'formNode' is missing in type 'typeof LoginForm'.
```

---

Using `typeof` to remove internal `formNode` from check:

``` ts
@muiThemeable<typeof LoginForm, LoginFormProps, LoginFormState>()
export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {}
```

Gives:

```
error TS2344: Type 'typeof LoginForm' does not satisfy the constraint 'Component<LoginFormProps, LoginFormState>'.
  Property 'setState' is missing in type 'typeof LoginForm'.
```

---

I've only got it working with `any`s:

``` ts
@muiThemeable<any, any, any>()
export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {}
```
## Solution?

I got it working by changing the definition to this:

``` ts
// material-ui.d.ts
export function muiThemeable<TComponent extends (typeof React.Component)>(): (component: TComponent) => TComponent;

// LoginForm.tsx
@muiThemeable<typeof LoginForm>()
export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {}
```

@ngbrown @herrmanno Thoughts? How did you get it working?

--- a/material-ui/index.d.ts
+++ b/material-ui/index.d.ts
@@ -452,7 +452,10 @@ declare namespace __MaterialUI {
         var lightBaseTheme: RawTheme;
         var darkBaseTheme: RawTheme;
 
-        export function muiThemeable<TComponent extends React.Component<P, S>, P, S>(): (component: TComponent) => TComponent;
+        export function muiThemeable(): <
+            TComponent extends React.ComponentClass<P> | React.StatelessComponent<P>,
+            P extends {muiTheme?: MuiTheme}
+        >(component: TComponent) => TComponent;
 
         interface MuiThemeProviderProps {
             muiTheme?: Styles.MuiTheme;

--- a/material-ui/material-ui-tests.tsx
+++ b/material-ui/material-ui-tests.tsx
@@ -7,6 +7,7 @@ import * as React from 'react';
 import {Component, PropTypes} from 'react';
 import * as ReactDOM from 'react-dom';
 import getMuiTheme from 'material-ui/styles/getMuiTheme';
+import {muiThemeable} from 'material-ui/styles/muiThemeable';
 import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
 import darkBaseTheme from 'material-ui/styles/baseThemes/darkBaseTheme';
 import {MuiTheme} from 'material-ui/styles';
@@ -319,6 +320,35 @@ class DeepDownTheTree extends React.Component<{} & {muiTheme: MuiTheme}, {}> {
     );
   }
 }
+
+
+const MuiThemeableFunction = muiThemeable()((props: {label: string, muiTheme?: MuiTheme}) => {
+  return (
+      <span style={{color: props.muiTheme.palette.textColor}}>
+        Applied the Theme to functional component: {props.label}.
+      </span>
+  );
+});
+
+@muiThemeable()
+class MuiThemeableClass extends React.Component<{label: string} & {muiTheme?: MuiTheme}, {}> {
+  render() {
+    return (
+      <span style={{color: this.props.muiTheme.palette.textColor}}>
+        Applied the Theme to class component decorated: {this.props.label}.
+      </span>
+    );
+  }
+}
+
+const MuiThemeableContainer = (props: {}) => (
+  <MuiThemeProvider muiTheme={getMuiTheme()}>
+    <div>
+      <MuiThemeableFunction label='Hello'/>
+      <MuiThemeableClass label='Hello'/>
+    </div>
+  </MuiThemeProvider>
+);
 
 
 // "http://www.material-ui.com/#/customization/inline-styles"

